### PR TITLE
Migrate gather into ATen

### LINKED ISSUE #24702: Migrate `gather` from the TH to Aten (CPU)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Fixes: https://github.com/pytorch/pytorch/issues/24702 #24567 

This prepares for implementing broadcasting in the future. This diff is mostly copy paste and variable renaming. 

Edit: remove the outdated benchmark

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
Also fixes #24567 (please assign both to yourself)

**Comment by VitalyFedyunin** [comment]
Can you please comment old TH code with note why it is not (yet) deleted.

**Comment by zasdfgbnm** [comment]
Also, is it a good idea to extend `TensorIterator` and use it? This sounds good to me, especially I need to implement broadcasting.

File: aten/src/ATen/native/ScatterGather.cpp [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    std::vector<int64_t> counter(num_dims, 0);
+
+    int64_t result_dim_stride = result.stride(dim);
+    int64_t index_dim_stride = index.stride(dim);
+    int64_t self_dim_stride = self.stride(dim);
+
+    while(!finished) {
+      for(int64_t j = 0; j < elems_per_row; j++) {
+        int64_t index_value = *(index_data + j * index_dim_stride);
+        AT_CHECK(index_value >= 0 && index_value < self_dim_size, "Invalid index in gather: out of range");
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/ScatterGather.cpp]: `TORCH_CHECK`  please

**Comment by VitalyFedyunin** [comment]
> Also, is it a good idea to extend `TensorIterator` and use it? This sounds good to me, especially I need to implement broadcasting.

Yes. It would be really nice to have DIM_APPLY loop written for TensorIterator!

File: aten/src/THC/generic/THCTensorScatterGather.cu [hunk 2]
```diff
@@ -9,6 +9,11 @@
 
 void THCTensor_(gather)(THCState* state, THCTensor *tensor,
                          THCTensor *src, int dim, THCudaLongTensor *index) {
+  // TODO (@zasdfgbnm): remove this.
```
Comment (gchanan) [inline aten/src/THC/generic/THCTensorScatterGather.cu]: is it used in the inner loop?  If not, you can call the aten function if you wrapper the THCTensors via THTensor_wrap.

Comment (zasdfgbnm) [reply aten/src/THC/generic/THCTensorScatterGather.cu]: Thanks for the information. Now I removed the legacy implementation using the `THTensor_wrap`

File: aten/src/ATen/Dispatch.h [hunk 3]
```diff
@@ -317,6 +317,27 @@ inline void deprecated_AT_DISPATCH_ALL_TYPES_AND_HALF_AND_COMPLEX() {}
     }                                                                                                            \
   }()
 
+#define AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND2(SCALARTYPE1, SCALARTYPE2, TYPE, NAME, ...)            \
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply aten/src/ATen/Dispatch.h]: @gchanan I think you are right. I have reverted this change

**Comment by VitalyFedyunin** [comment]
Hi! I'm doing final review, meanwhile, can you please do benchmarks for smaller/larger N, we need do decide if such performance change in acceptable.

**Comment by zasdfgbnm** [comment]
@VitalyFedyunin I have updated the benchmark code and result in the PR description. In summary, for large tensors, the time before and are after are roughly the same. For small tensors, the CPU code improves, but CUDA becomes slower. I guess this is caused not by the implementation itself, but by the difference in setting up things.

**Comment by zasdfgbnm** [comment]
@VitalyFedyunin I just did some profiling and found that the performance issue of this PR is mainly caused by an unnecessary memcpy. I have removed it in https://github.com/pytorch/pytorch/pull/24858/commits/3908dbe2c36af0dd213454de4778ee464c421278, the performance is now very close. So there is no performance issue to worry about now. Please do a last review :)

**Comment by VitalyFedyunin** [comment]
Hi! I will review today/tomorrow, meanwhile can you please rebase to master.

**Comment by zasdfgbnm** [comment]
@VitalyFedyunin rebased

**Comment by zasdfgbnm** [comment]
CI is broken

**Pull request closed without merging**